modules/tinymce/src/core/main/ts/api/html/DomParser.ts

```diff
@@ -362,6 +218,131 @@ const walker = (root: AstNode, settings: DomParserSettings, schema: Schema, node
   return state;
 };
 
+// All the dom operations we want to perform, regardless of whether we're trying to properly validate things
+// e.g. removing excess whitespace
+// e.g. removing empty nodes (or padding them with <br>)
+// e.g. handling data-mce-bogus
+const simplifyDom = (root: AstNode, schema: Schema, settings: DomParserSettings, args: ParserArgs) => {
+  const nonEmptyElements = schema.getNonEmptyElements();
+  const whitespaceElements = schema.getWhiteSpaceElements();
+  const blockElements: Record<string, string> = extend(makeMap('script,style,head,html,body,title,meta,param'), schema.getBlockElements());
+  const allWhiteSpaceRegExp = /[ \t\r\n]+/g;
+  const startWhiteSpaceRegExp = /^[ \t\r\n]+/;
+  const endWhiteSpaceRegExp = /[ \t\r\n]+$/;
+
+  const hasWhitespaceParent = (node: AstNode) => {
+    if (Obj.has(whitespaceElements, node.name)) {
+      return true;
+    } else if (node.parent) {
+      return hasWhitespaceParent(node.parent);
```

Comment:
This is potentially a lot of recursion which isn't great for a performance critical bit of code... it would probably be better to just use a for or while loop instead to walk up the parent.